tls/client-tls-pkcallback.c

```diff
@@ -0,0 +1,367 @@
+ */
+
+/* This example shows how to register a private key callback and optionally
+ * use the asynchronous version of the code to return a WC_PENDING_E while the
+ * hardware is processing */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <unistd.h>
+
+/* wolfSSL */
+#ifndef WOLFSSL_USER_SETTINGS
+    #include <wolfssl/options.h>
+#endif
+#include <wolfssl/ssl.h>
+#include <wolfssl/wolfcrypt/sha256.h>
+#include <wolfssl/wolfcrypt/cryptocb.h>
+#include <wolfssl/wolfcrypt/ecc.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+
+#define DEFAULT_PORT 11111
+
+#define USE_ECDHE_ECDSA
+#define USE_TLSV13
+
+#ifdef USE_ECDHE_ECDSA
+#define CERT_FILE "../certs/client-ecc-cert.pem"
+#define KEY_FILE  "../certs/ecc-client-key.pem"
+#define CA_FILE   "../certs/ca-ecc-cert.pem"
+#else
+#define CERT_FILE "../certs/client-cert.pem"
+#define KEY_FILE  "../certs/client-key.pem"
+#define CA_FILE   "../certs/ca-cert.pem"
+#endif
+
+typedef struct {
+    const char* keyFile;
+    ecc_key     key;
+    int         state;
+} PkCbInfo;
+
+#ifdef HAVE_PK_CALLBACKS
+/* reads file size, allocates buffer, reads into buffer, returns buffer */
+static int load_file(const char* fname, byte** buf, size_t* bufLen)
+{
+    int ret;
+    long int fileSz;
+    XFILE lFile;
+
+    if (fname == NULL || buf == NULL || bufLen == NULL)
+        return BAD_FUNC_ARG;
+
+    /* set defaults */
+    *buf = NULL;
+    *bufLen = 0;
+
+    /* open file (read-only binary) */
+    lFile = XFOPEN(fname, "rb");
+    if (!lFile) {
+        printf("Error loading %s\n", fname);
+        return BAD_PATH_ERROR;
+    }
+
+    fseek(lFile, 0, SEEK_END);
+    fileSz = (int)ftell(lFile);
+    rewind(lFile);
+    if (fileSz  > 0) {
+        *bufLen = (size_t)fileSz;
+        *buf = (byte*)malloc(*bufLen);
+        if (*buf == NULL) {
+            ret = MEMORY_E;
+            printf("Error allocating %lu bytes\n", (unsigned long)*bufLen);
+        }
+        else {
+            size_t readLen = fread(*buf, *bufLen, 1, lFile);
+
+            /* check response code */
+            ret = (readLen > 0) ? 0 : -1;
+        }
+    }
+    else {
+        ret = BUFFER_E;
+    }
+    fclose(lFile);
+
+    return ret;
+}
+
+static int load_key_file(const char* fname, byte** derBuf, word32* derLen)
+{
+    int ret;
+    byte* buf = NULL;
+    size_t bufLen;
+
+    ret = load_file(fname, &buf, &bufLen);
+    if (ret != 0)
+        return ret;
+
+    *derBuf = (byte*)malloc(bufLen);
+    if (*derBuf == NULL) {
+        free(buf);
+        return MEMORY_E;
+    }
+
+    ret = wc_KeyPemToDer(buf, (word32)bufLen, *derBuf, (word32)bufLen, NULL);
+    if (ret < 0) {
+        free(buf);
+        free(*derBuf);
+        return ret;
+    }
+    *derLen = ret;
+    free(buf);
+
+    return 0;
+}
+
+static int myEccSign(WOLFSSL* ssl, const byte* in, word32 inSz,
```

Comment:
Might want to add a comment that this function is simulating what hardware would be doing.

Reply:
Done for both